lang/funcs/defaults_test.go

```diff
@@ -0,0 +1,369 @@
+			}),
+			Defaults: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.StringVal("hello"),
+			}),
+			Want: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.StringVal("hey"),
+			}),
+		},
+		{
+			Input: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.NullVal(cty.String),
+			}),
+			Defaults: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.NullVal(cty.String),
+			}),
+			Want: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.NullVal(cty.String),
+			}),
+		},
+		{
+			Input: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.NullVal(cty.String),
+			}),
+			Defaults: cty.ObjectVal(map[string]cty.Value{}),
+			Want: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.NullVal(cty.String),
+			}),
+		},
+		{
+			Input: cty.ObjectVal(map[string]cty.Value{}),
+			Defaults: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.NullVal(cty.String),
+			}),
+			WantErr: `.a: target type does not expect an attribute named "a"`,
+		},
+
+		{
+			Input: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.ListVal([]cty.Value{
+					cty.NullVal(cty.String),
+				}),
+			}),
+			Defaults: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.StringVal("hello"),
+			}),
+			Want: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.ListVal([]cty.Value{
+					cty.StringVal("hello"),
+				}),
+			}),
+		},
+		{
+			Input: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.ListVal([]cty.Value{
+					cty.NullVal(cty.String),
+					cty.StringVal("hey"),
+					cty.NullVal(cty.String),
+				}),
+			}),
+			Defaults: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.StringVal("hello"),
+			}),
+			Want: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.ListVal([]cty.Value{
+					cty.StringVal("hello"),
+					cty.StringVal("hey"),
+					cty.StringVal("hello"),
+				}),
+			}),
+		},
+		{
+			// Using defaults with single set elements is a pretty
+			// odd thing to do, but this behavior is just here because
+			// it generalizes from how we handle collections. It's
+			// tested only to ensure it doesn't change accidentally
+			// in future.
+			Input: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.SetVal([]cty.Value{
+					cty.NullVal(cty.String),
+					cty.StringVal("hey"),
+				}),
+			}),
+			Defaults: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.StringVal("hello"),
+			}),
+			Want: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.SetVal([]cty.Value{
+					cty.StringVal("hey"),
+					cty.StringVal("hello"),
+				}),
+			}),
+		},
+		{
+			Input: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.MapVal(map[string]cty.Value{
+					"x": cty.NullVal(cty.String),
+					"y": cty.StringVal("hey"),
+					"z": cty.NullVal(cty.String),
+				}),
+			}),
+			Defaults: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.StringVal("hello"),
+			}),
+			Want: cty.ObjectVal(map[string]cty.Value{
+				"a": cty.MapVal(map[string]cty.Value{
+					"x": cty.StringVal("hello"),
+					"y": cty.StringVal("hey"),
+					"z": cty.StringVal("hello"),
+				}),
+			}),
```

Comment:
On first read I was confused about why we apply a single default value to each value of a map. Reading this test left me more confused, because I can't see how it would be useful.

It only made sense to me after reading the (excellent!) docs example, with a `map(object)`. It might be nice to add a test case similar to the example to clarify for anyone reading the code in future.

Reply:
I was trying to keep most of these tests focused on testing a single mechanism, and so this one is trying to test the specific situation of visiting each member of a map, alongside the adjacent ones testing visiting each member of a list, and each member of a set. There is a test further down that deals with recursively visiting the elements of an object in a collection, but I arbitrarily chose a list as the collection container for that test because the container part wasn't the focus of that test.

With that said, I suppose it wouldn't hurt to duplicate the list of objects test and turn it into a list of maps test with the same content, so I'll do that.

Reply:
As mentioned in the previous comment, I responded to this by adding a new test similar to the existing one with a list of objects (I actually took the one that was testing a _set_ of objects as the basis, but... same idea). I've retained the test this comment is attached to because I like having a test that is _only_ concerned with the "visit each element of a collection" part of the process so that the failures can help narrow down potential regressions in future maintenance.